plugin/keyring_vault/vault_keyring.cc

```diff
@@ -0,0 +1,198 @@
+  check_keyring_file_data,                                     /* check()    */
+  update_keyring_file_data,                                    /* update()   */
+  ""                                                           /* default    */
+);
+
+static struct st_mysql_sys_var *keyring_vault_system_variables[]= {
+  MYSQL_SYSVAR(config),
+  NULL
+};
+
+static int keyring_vault_init(MYSQL_PLUGIN plugin_info)
+{
+  try
+  {
+#ifdef HAVE_PSI_INTERFACE
+    keyring_init_psi_keys();
+#endif
+    SSL_library_init();
+    OpenSSL_add_all_algorithms();
+    SSL_load_error_strings();
+    curl_global_init(CURL_GLOBAL_NOTHING);
+
+    if (init_keyring_locks())
+      return TRUE;
+
+    logger.reset(new Logger(plugin_info));
+    keys.reset(new Vault_keys_container(logger.get()));
+    IVault_curl *vault_curl = new Vault_curl(logger.get());
+    IVault_parser *vault_parser = new Vault_parser(logger.get());
+    IKeyring_io *keyring_io= new Vault_io(logger.get(), vault_curl, vault_parser);
+    if (keys->init(keyring_io, keyring_vault_config_file))
+    {
+      is_keys_container_initialized = FALSE;
+      logger->log(MY_ERROR_LEVEL, "keyring_vault initialization failure. Please check that"
+        " the keyring_vault_config_file points to readable keyring_vault configuration"
+        " file. Please also make sure Vault is running and accessible."
+        " The keyring_vault will stay unusable until correct configuration file gets"
+        " provided.");
+      return FALSE;
+    }
+    is_keys_container_initialized = TRUE;
+    return FALSE;
+  }
+  catch (...)
+  {
+    if (logger != NULL)
+      logger->log(MY_ERROR_LEVEL, "keyring_vault initialization failure due to internal"
+                                  " exception inside the plugin");
+    return TRUE;
+  }
+}
+
+int keyring_vault_deinit(void *arg MY_ATTRIBUTE((unused)))
+{
+  //not taking a lock here as the calls to keyring_deinit are serialized by
+  //the plugin framework
+  ERR_remove_thread_state(NULL);
+  ERR_remove_state(0);
+  ERR_free_strings();
+  EVP_cleanup();
+
+  keys.reset();
+  logger.reset();
+  keyring_file_data.reset();
+  mysql_rwlock_destroy(&LOCK_keyring);
+
+  curl_global_cleanup();
+  return 0;
+}
+
+my_bool mysql_key_fetch(const char *key_id, char **key_type, const char *user_id,
+                        void **key, size_t *key_len)
+{
+  return mysql_key_fetch<keyring::Vault_key>(key_id, key_type, user_id, key, key_len);
+}
+
+my_bool mysql_key_store(const char *key_id, const char *key_type,
+                        const char *user_id, const void *key, size_t key_len)
+{
+  return mysql_key_store<keyring::Vault_key>(key_id, key_type, user_id, key, key_len);
+}
+
+my_bool mysql_key_remove(const char *key_id, const char *user_id)
+{
+  return mysql_key_remove<keyring::Vault_key>(key_id, user_id);
+}
+
+
+my_bool mysql_key_generate(const char *key_id, const char *key_type,
+                           const char *user_id, size_t key_len)
+{
+  try
```

Comment:
catch is here just in case. We did not want in any case for any exception to leave plugin and go into server - and break it, possibly causing a core dump. This is why those try catch function on plugin service functions.